x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts

```diff
@@ -632,6 +653,167 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
+    } catch (error) {
+      objects.forEach(({ type, id }) =>
+        this.auditLogger.log(
+          savedObjectEvent({
+            action: SavedObjectAction.COLLECT_MULTINAMESPACE_REFERENCES,
+            savedObject: { type, id },
+            error,
+          })
+        )
+      );
+      throw error;
+    }
+
+    // The user is authorized to access all of the requested objects in the space(s) that they exist in.
+    // Now: 1. omit any result objects that the user has no access to, 2. for the rest, redact any space(s) that the user is not authorized
+    // for, and 3. create audit records for any objects that will be returned to the user.
+    const requestedObjectsSet = objects.reduce(
+      (acc, { type, id }) => acc.add(`${type}:${id}`),
+      new Set<string>()
+    );
+    const filteredAndRedactedObjectsMap = response.objects.reduce(
+      (acc, { type, id, spaces, spacesWithMatchingAliases, inboundReferences }) => {
+        if (!spaces.length) {
+          return acc;
+        }
+        // Is the user authorized to access this object in all required space(s)?
+        const isAuthorizedForObject = isAuthorizedForObjectInAllSpaces(
+          type,
+          authAction,
+          typeActionMap,
+          [currentSpaceId]
+        );
+        // If the user is not authorized to access at least one inbound reference of this object, then we should omit this object.
+        // Note: this check relies on the fact that the order of the objects array is retained!
+        const isAuthorizedForInboundReference =
+          requestedObjectsSet.has(`${type}:${id}`) || // If true, this is one of the requested objects, and we checked authorization above
+          inboundReferences.some((inbound) => acc.has(`${inbound.type}:${inbound.id}`)); // If true, this object can be linked back to one of the requested objects
+
+        if (isAuthorizedForObject && isAuthorizedForInboundReference) {
+          this.auditLogger.log(
+            savedObjectEvent({
+              action: SavedObjectAction.COLLECT_MULTINAMESPACE_REFERENCES,
+              savedObject: { type, id },
+            })
+          );
+          const redactedSpaces = getRedactedSpaces(type, 'bulk_get', typeActionMap, spaces);
+          const redactedSpacesWithMatchingAliases =
+            spacesWithMatchingAliases &&
+            getRedactedSpaces(type, 'bulk_get', typeActionMap, spacesWithMatchingAliases);
+          return acc.set(`${type}:${id}`, {
+            type,
+            id,
+            spaces: redactedSpaces,
+            ...(redactedSpacesWithMatchingAliases && {
+              spacesWithMatchingAliases: redactedSpacesWithMatchingAliases,
+            }),
+            inboundReferences,
+          });
+        }
+        return acc;
+      },
+      new Map<string, SavedObjectReferenceWithContext>()
+    );
+
+    return {
+      objects: Array.from(filteredAndRedactedObjectsMap.values()),
```

Comment:
*thinking out loud*

The primary use case for this function is to collect the object graph so that we can subsequently share the entire graph to one or more spaces.

We are correctly filtering out objects that the user isn't authorized to see, but that will by definition return a subset of the object graph. When we discussed sharing without references, we were leaning towards not allowing for this, and instead require sharing references. The reasoning is that we expect that an object will require all of its references in order to function properly when shared to the next space (except for tags).

If we only share a subset of the references, then we are IMO in the same position as sharing no references: a potentially broken state.

^^ If we agree with that (and feel free to object!), then I wonder if we need this filtering logic at all. Would it instead be appropriate to throw an authorization error if the user isn't authorized for any of the objects in the graph? If not, should we indicate in our response that the graph was filtered due to insufficient access?

Reply:
My opinion is that the plugin consumer is responsible for maintaining their objects and the objects' references, and the plugin consumer is responsible for granting access to referenced types. If a user somehow has access to an object but cannot access that object's references, that's the plugin's fault, but we shouldn't try to account for that (buggy) edge case.

Part of my rationale for this behavior is that it mirrors how export and copy-to-space work. If you _somehow_ have access to Dashboards in a given space, but you don't have access to a Visualization that the Dashboard references, we don't throw an error if we export-with-references or copy-to-space-with-references, we just ignore the Visualization that the user doesn't have access to. Right now that situation isn't possible, but it very well could be in the future with OLS.

We could change this behavior for `collectMultiNamespaceReferences` to throw 403 errors more often, but I don't think it's a good idea.

Reply:
I'm leaning towards not allowing to share if a subset can't be accessed. I think that's more user-friendly in the long run as allowing to share most likely would result in a broken state.

Reply:
From an API perspective, it makes sense. I'm not so sure it makes sense from an end-user perspective, though.

What we are suggesting here is `collectMultiNamespaceReferences` will throw a 403 error if a user tries to collect references for an object (or objects) and the user doesn't have access to one of those references. So, if a user is in the Saved Object Management screen and they click the "Edit shared spaces" action for an object, they will experience an error.

If we are going to show them an error we need to try to be descriptive so they can resolve that error. I'm not sure how we will manage that in the Security plugin though, because we don't have a deep understanding of _why_ the user doesn't have access to the reference or what they need to do to get access to that reference.

I understand this is a corner case, and It Shouldn't Happen For Normal Usage™, but I am concerned that we will be causing more confusion in the future when this is the only feature that attempts to enforce referential integrity in this way.

Reply:
I chatted with @kobelb about this today and he raised a good point --

If a user was able to create a private visualization and add it as a reference to an existing dashboard, they could effectively deadlock the dashboard and prevent from being shared (because nobody else would have access to that visualization). That doesn't sound like a good user experience.

So I think we are better off treating these situations as "missing references" and potentially showing a warning in the UI, but allowing them to continue sharing the object(s). Does that sound reasonable to y'all?

Reply:
> So I think we are better off treating these situations as "missing references" and potentially showing a warning in the UI, but allowing them to continue sharing the object(s). Does that sound reasonable to y'all?

That sounds reasonable to me